File: rust/examples/ws_ticker_and_prices.rs
mod common;

use common::{market_and_orderbook, other, rest_client, ExampleResult};
use futures_util::StreamExt;
use lightcone::domain::price_history::wire::PriceHistory;
use lightcone::prelude::*;
use tokio::time::{timeout_at, Duration, Instant};

#[tokio::main]
async fn main() -> ExampleResult {
    let client = rest_client()?;
    let (_, orderbook) = market_and_orderbook(&client).await?;
    let mut ws = client.ws_native();
    let orderbook_id = orderbook.orderbook_id.clone();
    let mut history = PriceHistoryState::new();

    ws.connect().await?;
    ws.subscribe(SubscribeParams::Ticker {
        orderbook_ids: vec![orderbook_id.clone()],
    })?;
    ws.subscribe(SubscribeParams::PriceHistory {
        orderbook_id: orderbook_id.clone(),
        resolution: Resolution::Minute1,
        include_ohlcv: false,
    })?;

    let mut hits = 0;
    {
        let events = ws.events();
        tokio::pin!(events);

        let deadline = Instant::now() + Duration::from_secs(30);
        while hits < 4 {
            let Ok(Some(event)) = timeout_at(deadline, events.next()).await else {
                println!("no more websocket data (timeout or stream ended)");
                break;
            };

            match event {
                WsEvent::Message(Kind::Ticker(ticker)) => {
                    println!(
                        "ticker: bid={:?} ask={:?} mid={:?}",
                        ticker.best_bid, ticker.best_ask, ticker.mid
                    );
                    hits += 1;
                }
                WsEvent::Message(Kind::PriceHistory(PriceHistory::Snapshot(snapshot))) => {
                    history.apply_snapshot(
                        snapshot.orderbook_id.clone(),
                        snapshot.resolution,
                        snapshot.prices.into_iter().map(Into::into).collect(),
                    );
                    let candles = history
                        .get(&orderbook_id, &Resolution::Minute1)
                        .map_or(0, |points| points.len());
                    println!("price snapshot: {candles} candle(s)");
                    hits += 1;
                }
                WsEvent::Message(Kind::PriceHistory(PriceHistory::Update(update))) => {
                    history.apply_update(
                        update.orderbook_id.clone(),
                        update.resolution,
                        LineData {
                            time: update.t,
                            value: update.m.unwrap_or_default(),
                        },
                    );
                    println!(
                        "latest candle: {:?}",
                        history
                            .get(&orderbook_id, &Resolution::Minute1)
                            .and_then(|points| points.last())
                    );
                    hits += 1;
                }
                WsEvent::Message(Kind::PriceHistory(PriceHistory::Heartbeat(heartbeat))) => {
                    println!("heartbeat: {}", heartbeat.server_time);
                    hits += 1;
                }
                WsEvent::Error(err) => eprintln!("ws error: {err}"),
                _ => {}
            }
        }
    }

    ws.disconnect().await?;
    if hits == 0 {
        return Err(other("received no websocket events — connection may be broken").into());
    }
    Ok(())
}

File: typescript/examples/ws_ticker_and_prices.ts
import { PriceHistoryState, Resolution, type LineData, type WsEvent } from "../src";
import { marketAndOrderbook, restClient, withTimeout } from "./common";

async function main() {
  const client = restClient();
  const [, orderbook] = await marketAndOrderbook(client);
  const orderbookId = orderbook.orderbookId;
  const resolution = Resolution.Minute1;

  const history = new PriceHistoryState();
  const ws = client.ws();
  let hits = 0;

  let resolveDone!: () => void;
  const done = new Promise<void>((resolve) => {
    resolveDone = resolve;
  });

  const unsubscribe = ws.on((event: WsEvent) => {
    if (event.type === "Message" && event.message.type === "ticker") {
      const ticker = event.message.data;
      console.log(`ticker: bid=${ticker.best_bid} ask=${ticker.best_ask} mid=${ticker.mid}`);
      hits += 1;
    } else if (
      event.type === "Message" &&
      event.message.type === "price_history" &&
      event.message.data.event_type === "snapshot"
    ) {
      const data = event.message.data;
      const prices: LineData[] = data.prices.map((c) => ({
        time: c.t,
        value: c.m ?? "0",
      }));
      history.applySnapshot(data.orderbook_id, data.resolution, prices);
      const candles = history.get(orderbookId, resolution)?.length ?? 0;
      console.log(`price snapshot: ${candles} candle(s)`);
      hits += 1;
    } else if (
      event.type === "Message" &&
      event.message.type === "price_history" &&
      event.message.data.event_type === "update"
    ) {
      const data = event.message.data;
      const point: LineData = {
        time: data.t,
        value: data.m ?? "0",
      };
      history.applyUpdate(data.orderbook_id, data.resolution, point);
      console.log("latest candle:", history.get(orderbookId, resolution)?.at(-1));
      hits += 1;
    } else if (
      event.type === "Message" &&
      event.message.type === "price_history" &&
      event.message.data.event_type === "heartbeat"
    ) {
      console.log(`heartbeat: ${event.message.data.server_time}`);
    } else if (event.type === "Error") {
      console.error("ws error:", event.error);
    }

    if (hits >= 4) {
      resolveDone();
    }
  });

  try {
    await ws.connect();
    ws.subscribe({ type: "ticker", orderbook_ids: [orderbookId] });
    ws.subscribe({
      type: "price_history",
      orderbook_id: orderbookId,
      resolution,
      include_ohlcv: false,
    });
    await withTimeout(done, 15_000, "timed out waiting for websocket data");
  } finally {
    unsubscribe();
    await ws.disconnect();
  }
}

main().catch(console.error);
